main/src/library/SetMultiMap.flix

```diff
@@ -0,0 +1,514 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+
+///
+/// The SetMultiMap type.
+///
+/// A SetMultiMap is a Map that allows multiple values for a key.
+/// Multiple values are stored in a Set so duplicates are not allowed.
+///
+/// Value and key types must be implement Order.
+///
+pub enum SetMultiMap[k, v] {
+    case SetMultiMap(Map[k, Set[v]])
+}
+
+
+instance Boxable[SetMultiMap[k, v]] with Boxable[k], Boxable[v]
+
+instance ToString[SetMultiMap[k, v]] with ToString[k], ToString[v] {
+    // Uses `dropRight` to remove the trailing ", "
+    pub def toString(m: SetMultiMap[k, v]): String =
+        let SetMultiMap(m1) = m;
+        let s = Map.foldRightWithKey((k, v, acc) -> "${k} => ${v}, ${acc}", "", m1);
+        "SetMultiMap#{${String.dropRight(2, s)}}"
+}
+
+instance Hash[SetMultiMap[k, v]] with Hash[k], Hash[v] {
+    pub def hash(m: SetMultiMap[k, v]): Int32 =
+        let SetMultiMap(m1) = m;
+        Map.foldLeftWithKey((acc, k, v) -> acc * 17 + Hash.hash(k) + 19 * Hash.hash(v), 7937, m1)
+}
+
+
+instance Eq[SetMultiMap[k, v]] with Eq[k], Eq[v] {
+    pub def eq(m1: SetMultiMap[k, v], m2: SetMultiMap[k, v]): Bool =
+        SetMultiMap.toList(m1) == SetMultiMap.toList(m2)
+}
+
+instance Order[SetMultiMap[k, v]] with Order[k], Order[v] {
+    pub def compare(x: SetMultiMap[k, v], y: SetMultiMap[k, v]): Comparison =
+        SetMultiMap.toList(x) <=> SetMultiMap.toList(y)
+}
+
+instance Foldable[SetMultiMap[k]] {
+    pub def foldLeft(f: (b, v) -> b & ef, s: b, m: SetMultiMap[k, v]): b & ef = SetMultiMap.foldLeft((acc, v) -> f(acc, v), s, m)
+    pub def foldRight(f: (v, b) -> b & ef, s: b, m: SetMultiMap[k, v]): b & ef = SetMultiMap.foldRight(f, s, m)
+    pub def foldRightWithCont(f: (v, Unit -> b & ef) -> b & ef, s: b, m: SetMultiMap[k, v]): b & ef = SetMultiMap.foldRightWithCont(f, s, m)
+}
+
+instance SemiGroup[SetMultiMap[k,v]] with Order[k], Order[v] {
+    pub def combine(x: SetMultiMap[k,v], y: SetMultiMap[k,v]): SetMultiMap[k,v] = SetMultiMap.union(x, y)
+}
+
+instance CommutativeSemiGroup[SetMultiMap[k,v]] with Order[k], Order[v]
+
+instance Monoid[SetMultiMap[k, v]] with Order[k], Order[v] {
+    pub def empty(): SetMultiMap[k, v] = SetMultiMap.empty()
+}
+
+instance CommutativeMonoid[SetMultiMap[k, v]] with Order[k], Order[v]
+
+namespace SetMultiMap {
+
+    ///
+    /// Returns the empty SetMultiMap.
+    ///
+    pub def empty(): SetMultiMap[k, v] = SetMultiMap(Map.empty())
+
+    ///
+    /// Returns the singleton SetMultiMap where key `k` is mapped to value `v`.
+    ///
+    pub def singleton(k: k, v: v): SetMultiMap[k, v] with Order[k], Order[v] = SetMultiMap(Map.singleton(k, Set.singleton(v)))
+
+    ///
+    /// Returns a SetMultiMap where key `k` is mapped to the singleton set containing `v`.
+    ///
+    pub def fromMap(m: Map[k, v]): SetMultiMap[k, v] with Order[k], Order[v] = SetMultiMap(Map.map(Set.singleton, m))
```

Comment:
We usually don't have `from`, so I would propose to remove this and instead add various `toMultiMap` functions.

Reply:
I mean, we never really resolved the from/to. We kind of just ended up deciding to focus on toXYZ because then you can browse a namespace to discover what your options are.

Reply:
I've removed `fromMap` and added `Map.toMultiMap`.

Also added `MultiMap.toMap` which I'd missed previously.